### PR TITLE
Add conv2d bwd data

### PR DESCRIPTION
**Pull request by jerryyin**
Continued from #1

### CODE REVIEW

File: miopen_dialect_idea.md [hunk 1]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
 
 ```mlir
 // apply gridwise GEMM
 miopen.gridwise_gemm(%filter_gemmK_gemmM, %input_gemmK_gemmN, %output_gemmM_gemmN) {
-  parameters = [
-    // tuning parameters
-  ]
+  filter_layout = ["k", "c", "y", "x"],
+  input_layout = ["n", "c", "hi", "wi"],
+  output_layout = ["n", "k", "ho", "wo"],
```
Comment (whchung) [inline miopen_dialect_idea.md:197]: missing other attributes here:
- dilations
- padding
- strides
- filter_dimension
- input_dimension
- output_dimension

File: miopen_dialect_idea.md [hunk 2]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+  ],
+  output_layout = ["gemmK", "gemmM"],
+  source_layout = ["k", "c", "y", "x"]
+} : memref<?x?x?x?xf32> to memref<?x?xf32>
+```
+
+```mlir
+// output_diff tensor
+%output_gemmM_gemmN = miopen.transform(%output_diff) {
+  gridwise_gemm_argument_position = 2
```
Comment (whchung) [inline miopen_dialect_idea.md]: gridwise_gemm_argument_position = 1?

File: miopen_dialect_idea.md [hunk 3]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+        source_dimensions = [2],
+        source_names = ["wipad"]
+      }
+  ],
+  intermediate_layout = ["n", "c", "hipad", "wipad"]
+  output_layout = ["n", "c", "y", "ho", "x", "wo"]
+} : memref<?x?x?x?xf32> to memref<?x?x?x?x?x?x?xf32>
+
+%input_gemmK_gemmN = miopen.transform(%input_n_c_y_ho_x_wo) {
+  layout = [
```
Comment (whchung) [inline miopen_dialect_idea.md:342]: gridwise_gemm_argument_position = 2?

File: miopen_dialect_idea.md [hunk 4]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+  ],
+  intermediate_layout = ["n", "c", "y", "ho", "x", "wo"],
+  output_layout = ["gemmK", "gemmN"]
+} : memref<?x?x?x?x?x?x?xf32> to memref<?x?xf32>
+```
+
+```mlir
+// apply gridwise GEMM
+miopen.gridwise_gemm(%filter_gemmK_gemmM, %output_gemmM_gemmN, %input_gemmK_gemmN) {
+  filter_layout = ["k", "c", "y", "x"],
```
Comment (whchung) [inline miopen_dialect_idea.md:368]: missing attributes

File: miopen_dialect_idea.md [hunk 5]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+    }
+  ],
+  intermediate_layout = ["n", "c", "y", "ho", "x", "wo"],
+  output_layout = ["gemmK", "gemmN"]
+} : memref<?x?x?x?x?x?x?xf32> to memref<?x?xf32>
+```
+
+```mlir
+// apply gridwise GEMM
+miopen.gridwise_gemm(%filter_gemmK_gemmM, %output_gemmM_gemmN, %input_gemmK_gemmN) {
```
Comment (whchung) [inline miopen_dialect_idea.md]: is it a valid matrix multiplication to do  transpose(K * M) * (M * N) ?

File: miopen_dialect_idea.md [hunk 6]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+  filter_layout = ["k", "c", "y", "x"],
+  input_layout = ["n", "c", "hi", "wi"],
+  output_layout = ["n", "k", "ho", "wo"],
+} : memref<?x?xf32>,
```
Comment (whchung) [inline miopen_dialect_idea.md:376]: how can the dialect be amended so `miopen.gridwise_gemm` can keep track the information so at MLIR -> C++ translation stage we know it's doing backward data, so a different algorithm v1r1 should be used?

**Review by whchung — CHANGES_REQUESTED** [review]
initial round of review.

File: miopen_dialect_idea.md [hunk 7]
```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
...
+An example based on NCHW/KCYX/NKHW:
+
+```mlir
+// filter tensor
+%filter_gemmK_gemmM = miopen.transform(%filter) {
+  gridwise_gemm_argument_position = 0
+  layout = [
+    {
+      dimensions = [0],
+      names = ["gemmK"],
```
Comment (whchung) [inline miopen_dialect_idea.md:219]: gemmK or gemmM?

File: miopen_dialect_idea.md (hunk 1 above)
Comment (jerryyin) [reply miopen_dialect_idea.md:197]: Added

File: miopen_dialect_idea.md (hunk 5 above)
Comment (jerryyin) [reply miopen_dialect_idea.md]: Updated to `transpose(K, M) * (K, N) = (M, N)`

File: miopen_dialect_idea.md (hunk 6 above)
Comment (jerryyin) [reply miopen_dialect_idea.md:376]: Added field `kernel_algorithm`;

In forward pass, `kernel_algorithm = v4r4`
In backward data pass, `kernel_algorithm = backward_data_v1r1`

I pick the algorithm name to be aligned with the MLOpen algorithm file.

Comment (jerryyin) [reply miopen_dialect_idea.md:376]: If we are on the same page with this field, my next step is to add the field to existing operation, and make sure parsing, lowering, translating continue to work as before. I will file a new PR for that.

**Review by whchung — APPROVED** [review]

**Pull request merged**